_posts/2019-04-10-aws-dashboard-console-using-arduino.md

```diff
@@ -0,0 +1,265 @@
+#define ALARMS_LED D0
+#define PIPE_RUNNING_LED_DEV D8
+#define PIPE_FAILED_LED_DEV D7
+#define PIPE_RUNNING_LED_TEST D6
+#define PIPE_FAILED_LED_TEST D5
+#define PIPE_RUNNING_LED_PROD D4
+#define PIPE_FAILED_LED_PROD D3
+
+void setup() {
+  Serial.begin(115200);
+  
+  pinMode(ALARMS_LED, OUTPUT);
+  pinMode(PIPE_RUNNING_LED_DEV, OUTPUT);
+  pinMode(PIPE_FAILED_LED_DEV, OUTPUT);
+  pinMode(PIPE_RUNNING_LED_TEST, OUTPUT);
+  pinMode(PIPE_FAILED_LED_TEST, OUTPUT);
+  pinMode(PIPE_RUNNING_LED_PROD, OUTPUT);
+  pinMode(PIPE_FAILED_LED_PROD, OUTPUT);
+
+  Serial.print("START");
+
+  lcd.init();
+  lcd.backlight();
+
+  initConsole();
+  connectWifi();
+  
+  delay(2000);
+  lcd.setCursor(0, 0);
+  lcd.print("Call AWS...     ");
+  lcd.setCursor(0, 1);
+  lcd.print("                ");
+}
+```
+
+First, the LED pins are defined. This definition is required to setup which IO pins the LED wires are connected. Then the IO pin mode is set to “OUTPUT” which is required to enable a write mode for the pins.
+
+The LCD screen is initialized and backlight is switch on. The electronic components are initialized by the custom method `initConsole`. Basically the `initConsole` will blink the LEDs and write some text to the screen. Finally, the WiFi is connected.
+
+**After the `setup` has finished, start the infinite `loop`:**
+
+```c++
+void loop() {
+  if (WiFi.status() == WL_CONNECTED) {
+    // Call the AWS Lambda endpoints
+    String data_dev = callAWS(host_dev, "D", api_key_dev);
+    String data_test = callAWS(host_test, "T", api_key_test);
+    String data_prod = callAWS(host_prod, "P", api_key_prod);
+    
+    // Set the console
+    setConsole(data_dev, data_test, data_prod);        
+  } else {
+    lcd.print("Wifi Conn lost   ");
+  }
+}
+```
+
+The each AWS environment (DEV, TEST and PROD) contains own Lambda script to serve status information. Environments are call separately and results are passed to the `setConsole` method that controls the electronics. 
+> *In this context the environment is equal with an AWS account.*
+
+**The HTTP request is made inside the `callAWS` method:**
+
+```c++
+ client.print(String("GET ") + url + " HTTP/1.1\r\n" +
+               "Host: " + host + "\r\n" +
+               "User-Agent: ArduinoRadiatorConsole\r\n" +
+               "x-api-key: " + api_key + "\r\n" +
+               "Connection: close\r\n\r\n");
+
+  Serial.println("request sent");
+
+  while (client.connected()) {
+    String line = client.readStringUntil('\n');
+
+    if (line == "\r") {
+      Serial.println("headers received");
+      break;
+    }
+  }
+
+  String line = client.readStringUntil('}');
+  line = line + "}";
+
+  client.stop();
+```
+
+The low level library `WiFiClientSecure` is used for the HTTP request. High level HTTP libraries are also available but I wanted to stay in basics and decided to use this low level implementation. The HTTP response is read until a closing parenthesis which is a last character of the HTTP REST response.
```

Comment:
until closing parenthesis is encountered tms.

Reply:
The low level  --> The low-level